sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala

```diff
@@ -342,11 +343,49 @@ object FileStreamSource {
     def size: Int = map.size()
   }
 
-  private[sql] trait FileStreamSourceCleaner {
-    def clean(entry: FileEntry): Unit
+  private[sql] abstract class FileStreamSourceCleaner(
+      val fileSystem: FileSystem,
+      val sourcePath: Path) extends Logging {
+
+    private val srcPathToContainFileStreamSinkMetadata = new mutable.HashMap[Path, Boolean]
+
+    def clean(entries: Seq[FileEntry]): Unit = {
+      doClean(excludeFilesFromFileStreamSink(entries))
+    }
+
+    private def excludeFilesFromFileStreamSink(entries: Seq[FileEntry]): Seq[FileEntry] = {
+      val srcPathToEntries = entries.groupBy { entry =>
+        // find the ancestor which has same depth as source path
+        var curPath = new Path(new URI(entry.path))
+        while (curPath.depth() > sourcePath.depth()) {
+          curPath = curPath.getParent
+        }
+        curPath
+      }
+
+      srcPathToEntries.filterKeys { srcPath =>
+        srcPathToContainFileStreamSinkMetadata.get(srcPath) match {
+          case Some(v) => !v
```

Comment:
+1 on caching since we spare quite a time.
Not yet sure whether it's forbidden to set existing directory as sink (haven't found any explicit statement)?
If it's allowed then cache would contain `false` because `/a/b/c/d` file found but no `_spark_metadata`. All of a sudden a sink query started on `/a/b/c` which makes cached value invalid.

Reply:
Yeah that's the reason I asked for more voices - the value of cache can be invalid at any time, but then we can't cache it and have to check every time which is resource-inefficient. Maybe I'd be even OK to not use cache given we'll do it in background, but want to check if it's only me.